[1단계 - 출석 구현] 젠슨 미션 제출합니다.

CodingMasterLSW opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
 
https://prolog.techcourse.co.kr/studylogs/3959

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
어느정도 객체를 나눴지만, 구현에 집중해 개행 및 depth 등에 대한 리팩토링을 진행하지 못 했음.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

객체지향 생활체조를 지키지 못 했고, 이는 충분히 스스로 리팩토링 할 수 있지 않나 싶습니다. (시간이 없어 리팩토링을 하지는 못 했습니다)

전반적으로 객체의 책임 분리, 설계에 대해 리뷰해주시면 감사하겠습니다!


<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

1) 값 수정 메서드에 대해서
```
public void modify(LocalTime localTime, AttendanceType attendanceType) {
        this.attendanceTime = LocalDateTime.of(this.attendanceTime.toLocalDate(), localTime);
        this.attendanceType = attendanceType;
    }
```
modify() 로직을 수행할 때, 필드값을 바꾸는 게 좋을지, 새로운 객체를 생성할 지 궁금합니다.
그렇게 생각한 이유 : 객체 생성 비용이 있어 필드값을 수정하는게 좋다고 판단했지만, 객체를 새로 생성해도 관리되지 않는 객체는 GC 대상이기에 알아서 관리해줘서 이 방법도 괜찮겠다 생각이 들었습니다.

2) view의 책임은 어디까지?
```
private void sortDangerousCrews(LocalDate now, List<Crew> dangerousCrews) {
        dangerousCrews.sort(new Comparator<Crew>() {
            @Override
            public int compare(Crew o1, Crew o2) {
                CrewStatus crewStatus1 = o1.calculateCrewStatus(o1.calculateAttendanceResult(now));
                CrewStatus crewStatus2 = o2.calculateCrewStatus(o2.calculateAttendanceResult(now));
                if (crewStatus1 == crewStatus2) {
                    return o1.getName().compareTo(o2.getName());
                }
                return crewStatus2.getOrder() - crewStatus1.getOrder();
            }
        });
    }
```
요구사항대로 값을 정렬하기 위해 view에서 comparator를 정의했습니다.  

출력을 위한 정렬이라고 생각했고, view에서 해당 로직을 담당하도록 설계했습니다.
제가 생각하는 view는 복잡한 로직 없이 출력을 담당하는 역할이라고 생각하는데, 위의 코드처럼 view에서 정렬 로직을 정의하는게 맞는지 궁금합니다.

3) TDD 관련 질문
이번 미션에서 TDD를 진행하면서 메서드에 관련해서는 tdd를 어떻게 사용해야 하는지 어느정도 알 수 있었습니다.
[385a6c8](https://github.com/woowacourse/java-attendance/pull/22/commits/385a6c899d9184f809dcbadda6fec9e5cf14f60e)
하지만 객체를 만드는 과정에서부터 tdd가 좀 꼬였다고 느꼈습니다.
위의 코드처럼 객체를 만드는 과정, 즉 객체 생성에 대한 부분도 tdd를 도입하는게 좋다고 생각하시나요? 또한, 위의 커밋처럼 객체 생성에 대한 테스트 코드 또한 필수적이라고 생각하시는지 궁금합니다.

kjyyjk committed:
docs: 초기 기능명세서 작성

kjyyjk committed:
docs: 출석 확인 세부 기능명세 작성

kjyyjk committed:
feat: 등교 시간을 토대로 출석, 지각, 결석 판별 구현

kjyyjk committed:
feat: 주어진 날짜를 바탕으로 공휴일 혹은 주말 판별

kjyyjk committed:
feat: 출석 결과를 저장

kjyyjk committed:
feat: 출석 수정 로직 구현

kjyyjk committed:
docs: 출석 확인 및 출석 수정 세부 기능명세 작성

kjyyjk committed:
rename: 출석 관련 클래스 및 변수명 수정

kjyyjk committed:
feat: 중복 출석 시 검증 구현

- Crew 객체의 책임을 명확히 하기 위해 AttendancHistory를 도입하였음

kjyyjk committed:
feat: 주말 또는 공휴일에 출석할 경우 예외 처리

kjyyjk committed:
feat: 출석 기록이 존재하지 않는다면 예외 발생

kjyyjk committed:
refactor: AttendanceType Enum 상수로 변경

kjyyjk committed:
feat: 기능 선택 구현

- 존재하지 않는 기능일 경우, 예외를 발생시킨다.

kjyyjk committed:
feat: 출석 확인 기능 구현

kjyyjk committed:
rename: ABSENT에서 ABSENCE로 상수 이름 변경

- ENUM 상수명을 명사형으로 통일하기 위해 변경하였다.

kjyyjk committed:
feat: 출석 수정 기능 구현

kjyyjk committed:
rename: 클래스명 수정

kjyyjk committed:
refactor: 출석 수정 로직 변경

kjyyjk committed:
feat: 크루 별 출석 기록 확인 기능 구현

kjyyjk committed:
feat: 출석, 지각, 결석 횟수 출력 기능 구현

kjyyjk committed:
feat: 제적 위험자 확인 후 출력

kjyyjk committed:
refactor: 시스템의 오늘 날짜를 2024년 12월로 변경

kjyyjk committed:
refactor: Controller의 메서드 분리

kjyyjk committed:
refactor: 기능 단위 별 메서드 분리

kjyyjk committed:
feat: 파일 읽은 후 크루 및 출석 기록 초기화

kjyyjk committed:
refactor: 유틸 클래스 private 생성자로 변경

kjyyjk committed:
docs: 크루 출석 기록 추가

kjyyjk committed:
feat: 캠퍼스 운영 시간 기준으로 출석 시간 검증 구현

kjyyjk committed:
feat: 출력 시 개행 추가

ddaaac commented on src/main/java/attendance/SystemCurrentDate.java:
```diff
@@ -0,0 +1,10 @@
+public class SystemCurrentDate implements CurrentDate {
+    @Override
+    public LocalDate now() {
+        return LocalDate.now().minusMonths(2);
```
어차피 3월 넘어가면 - 3 해야되는거라면 그냥 여기를 12월 30일 이렇게 박아넣어도 괜찮아보여요. 12월 날짜를 사용하는게 요구사항이기도 하구요.

CodingMasterLSW replied:
수정해서 반영해보겠습니다. 
이 당시에 위와 같이 구현한 이유는, 현재 시간으로 3월이 되면, 요구사항이 12->1로 바뀌는게 자연스럽다고 생각해 아래와 같이 구현을 진행했었습니다.

ddaaac commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,134 @@
+                    return o1.getName().compareTo(o2.getName());
+                }
+                return crewStatus2.getOrder() - crewStatus1.getOrder();
+            }
```
> 출력을 위한 정렬이라고 생각했고, view에서 해당 로직을 담당하도록 설계했습니다.
제가 생각하는 view는 복잡한 로직 없이 출력을 담당하는 역할이라고 생각하는데, 위의 코드처럼 view에서 정렬 로직을 정의하는게 맞는지 궁금합니다.


이 로직이 잘 동작하는지 어떻게 테스트해봐야할까요? 이 로직에 대한 동작을 테스트하기 위해 View에 대한 테스트를 작성해야할까요?
또, TDD로 구현한다면 테스트가 먼저 작성돼야합니다.

정렬은 View와 상관없는 도메인 요구사항일 수 있습니다.(이번 요구사항에서도 도메인 요구사항으로 해석할만하구요) 도메인으로 옮겨서 테스트도 작성하면 좋을 것 같아요.

CodingMasterLSW replied:
좋은 의견 감사합니다. 수정해보겠습니다

ddaaac commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,61 @@
+        return checkGeneralAttendanceType(localTime);
+    }
+
+    public static void checkHoliday(LocalDate localDate) {
```
조금 직역해보면 "휴일임을 확인한다"로 해석할 수 있는데요, 그렇게 되면 하는 일과 반대의 네이밍이 되는 것 같네요.

CodingMasterLSW replied:
메서드 명을 조금 구체화 해보겠습니다!

ddaaac commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,134 @@
+        System.out.println(ATTENDANCE_RESULT_MESSAGE.formatted(
+                attendanceTime.getMonthValue(),
+                attendanceTime.getDayOfMonth(),
+                dayOfWeek,
```
한글로 보여야하지 않을까요?

CodingMasterLSW replied:
맞습니다. 페어 노트북에서 dayOfWeek.displayName() 에 대한 메서드 사용이 안 되었고, 문제 해결을 하지 못 해 수정하지 못 했습니다. 수정하도록 하겠습니다.

ddaaac commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,61 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public class AttendancePolicy {
```
모든 메서드가 static이네요.

요일을 나타내는 객체/클래스를 하나 두면 여기 있는 정책들을 좀 더 분명하게 표현할 수도 있을 것 같구요.

CodingMasterLSW replied:
역할을 조금 더 명확하게 분리해보겠습니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,45 @@
+
+    public AttendanceHistory(LocalDateTime attendanceTime, AttendanceType attendanceType) {
+        this.attendanceTime = attendanceTime;
+        this.attendanceType = attendanceType;
```
전 코멘트에서 말한 요일 나타내는 객체가 있다면 type도 여기서 계산할 수 있을 것 같네요.

ddaaac commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+        AttendanceHistory attendanceHistory = crew.getAttendanceHistory(modifyDate);
+        AttendanceHistoryDto beforeAttendanceHistoryDto = AttendanceHistoryDto.of(attendanceHistory);
+        AttendanceHistory afterAttendanceHistory = crew.modifyAttendanceResult(attendanceHistory, modifyTime);
+        outputView.printModifyAttendanceResult(beforeAttendanceHistoryDto, afterAttendanceHistory);
```
dto를 사용해주신 기준이 궁금합니다.

CodingMasterLSW replied:
날짜 변경 이후에 attendanceHistory에 대한 참조값이 연결되어 있어, 수정 이전의 결과를 저장할 필요가 있었습니다.
원본 날짜를 출력하기 위해서는 변경 전 날짜를 어딘가에 값을 저장해뒀어야 했고, 이 문제를 해결하기 위해 Dto에 값을 저장해서 사용했습니다.

ddaaac replied:
dto를 사용할거라면 모두 통일해서 사용하고, 사용하지 않을거라면 사용하지 않는 쪽으로 통일하면 좋을 것 같아요
잘 수정해주셨습니다 👍

ddaaac commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+    public Map<AttendanceType, Integer> calculateAttendanceResult(LocalDate localDate) {
+        Map<AttendanceType, Integer> attendanceResult = initializeAttendanceResult();
+        for (int i = 1; i < localDate.getDayOfMonth(); i++) {
+            LocalDate date = LocalDate.of(localDate.getYear(), localDate.getMonthValue(), i);
```
`localDate.withDayOfMonth(i)` 요런 것도 있습니다.

CodingMasterLSW replied:
수정해보도록 하겠습니다!

ddaaac commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+        for (int i = 1; i < localDate.getDayOfMonth(); i++) {
+            LocalDate date = LocalDate.of(localDate.getYear(), localDate.getMonthValue(), i);
+            try {
+                AttendancePolicy.checkHoliday(date);
```
checkHoliday랑 isHoliday를 구분해도 좋겠네요.
try ~ catch는 에러 핸들링에 사용해주세요. 로직은 다른 방식으로 풀구요.

CodingMasterLSW replied:
확인했습니다. 
예외 상황일 경우에만, try-catch를 사용하도록 로직 변경하겠습니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+            for (AttendanceHistory attendanceHistory : attendanceHistories) {
+                if (attendanceHistory.getAttendanceTime().toLocalDate().equals(date)) {
+                    AttendanceType attendanceType = attendanceHistory.getAttendanceType();
+                    attendanceResult.put(attendanceType, attendanceResult.get(attendanceType) + 1);
```
`computeIfPresent`를 활용해도 좋을 것 같아요.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+                    AttendanceType attendanceType = attendanceHistory.getAttendanceType();
+                    attendanceResult.put(attendanceType, attendanceResult.get(attendanceType) + 1);
+                    flag = true;
+                }
```
변수명, depth 등은 리뷰 반영하면서 변경해주시면 좋을 것 같습니다.

CodingMasterLSW replied:
넵 알겠습니다

ddaaac commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+        return modifyAttendanceHistory;
+    }
+
+    public Map<AttendanceType, Integer> calculateAttendanceResult(LocalDate localDate) {
```
테스트가 존재하지 않는 것 같아요. 테스트를 작성해주세요.

> 위의 코드처럼 객체를 만드는 과정, 즉 객체 생성에 대한 부분도 tdd를 도입하는게 좋다고 생각하시나요? 또한, 위의 커밋처럼 객체 생성에 대한 테스트 코드 또한 필수적이라고 생각하시는지 궁금합니다.


테스트가 없다는 건 TDD 사이클을 활용하지 않았다는 얘기 같네요.
TDD는 단순히 테스트의 생성 때문에 사용하는 방법이 아닙니다.
테스트를 먼저 작성함으로써 어떤 책임을 가진 객체가 필요한지, 그 객체의 책임을 어떻게 표현할지 등을 클라이언트 사이드에서 먼저 생각하는 방식이에요.
그래서 당연하게도 객체의 이름을 정하고, 어떤 파라미터를 넣어줄 지 등을 정하기 위해 테스트 코드를 먼저 작성해야하겠죠.

CodingMasterLSW replied:
미션 구현에 급급해 TDD를 초반에만 진행하다, 일반 구현으로 미션을 진행했었습니다.
다음 미션에서는 미션 구현을 다 하지 못 해도, TDD 적용을 우선적으로 생각하려고 합니다.

ddaaac replied:
오 아닙니다 ㅋㅋㅋ
언제나 구현이 먼저죠. 
2단계에서는 이미 도메인 지식이 있는 상태에서 처음부터 할 수 있으니 2단계에서 TDD 신경써서 해보시면 좋을 것 같습니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistory.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void modify(LocalTime localTime, AttendanceType attendanceType) {
+        this.attendanceTime = LocalDateTime.of(this.attendanceTime.toLocalDate(), localTime);
+        this.attendanceType = attendanceType;
```
> modify() 로직을 수행할 때, 필드값을 바꾸는 게 좋을지, 새로운 객체를 생성할 지 궁금합니다.
그렇게 생각한 이유 : 객체 생성 비용이 있어 필드값을 수정하는게 좋다고 판단했지만, 객체를 새로 생성해도 관리되지 않는 객체는 GC 대상이기에 알아서 관리해줘서 이 방법도 괜찮겠다 생각이 들었습니다.


구현하기 나름이긴 한데, 생성 비용 때문에 불변 객체를 사용하지 않을 필요는 없습니다.

[오라클 공식 문서](https://docs.oracle.com/javase/tutorial/essential/concurrency/immutable.html)를 보시면 이런 내용이 있습니다.

> Programmers are often reluctant to employ immutable objects, because they worry about the cost of creating a new object as opposed to updating an object in place. The impact of object creation is often overestimated, and can be offset by some of the efficiencies associated with immutable objects. These include decreased overhead due to garbage collection, and the elimination of code needed to protect mutable objects from corruption.


생성 비용을 제외한 각 방식의 장단점을 생각해보고 결정해주시면 될 것 같아요.

ddaaac reviewed (changes requested):
안녕하세요 젠슨 👋 

출석 미션 구현 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

CodingMasterLSW committed:
refactor: 운영 시간 enum으로 관리

CodingMasterLSW committed:
refactor: DayOfWeek에서 지각 시간 계산 및 AttendanceType 객체가 스스로 타입 반환 하도록 수정

CodingMasterLSW committed:
refactor: AttendancePolicy 로직 수정

Holiday, AttendanceType, DayOfWeek 각 객체에 역할을 부여하였음

CodingMasterLSW committed:
refactor: attendanceTime 원시값 포장 및 출석 기록 수정 로직 변경

AttendanceHistory의 책임을 보다 명확히 하기 위해 AttendanceTime을 객체로 포장하였음. 또한, 매개변수의 값을 바꾸는 수정 형식에서, 새로운 객체를 반환하는 형식으로 수정 로직 리팩토링을 진행하였음.

CodingMasterLSW committed:
refactor: 출석 수정 로직 변경

필드값을 수정하는 기존 로직에서 변경된 객체 삭제 후 재생성 하도록 비즈니스 로직을 수정하였음

CodingMasterLSW committed:
refactor: Crew와 AttendanceHistory 분리

객체의 책임을 명확히 하기 위해 Crew와 AttendanceHistory를 분리하고, CrewAttendanceManager를 도입하였음

CodingMasterLSW committed:
refactor: 변경사항에 알맞게 출석 수정 로직 돌아가도록 리팩토링 진행

CodingMasterLSW committed:
refactor: 출석 기록이 없는 경우, 결석 로직 구현

CodingMasterLSW committed:
refactor: 분리한 객체에 맞게 크루 별 출석 기록 확인, 제적 위험자 확인 로직 수정

CodingMasterLSW committed:
refactor: depth 조절 및 메서드 분리

CodingMasterLSW committed:
refactor: 제적위험자를 관리하는 dangerousCrews 생성 후 관리

CodingMasterLSW committed:
refacrtor: 불변객체 반환 및 메세지 상수화

CodingMasterLSW commented:
피드백 주신대로 객체의 책임을 조금 더 명확하게 나눠봤습니다!

1) 다만, crew 객체의 역할이 조금 모호하다고 생각해 crewStatus를 crew객체에 포함시킬까 고민중입니다.
이 경우에, crewStatus의 default 값을로 CrewStatus.CLEAR로 설정하고, 이후에 크루 상태를 변경하려고 합니다.
이런 리팩토링 사항에 대해 범블비의 의견이 궁금합니다!

2) 기존에는 view에서 제적위험자를 정렬하고 있었습니다. 이를 수정하기 위해 controller에서 정렬을 시도했었고, 
controller에서 비즈니스 로직을 수행하는건 올바르지 않다 생각해 DangerousCrew, DangerousCrews 객체 생성 후 책임을 전가했습니다.
범블비가 봤을 때, DangerousCrew는 필요한 객체라고 생각하시나요? 아니면 과한 분리라고 생각하시나요?? 의견이 궁금합니다.

3) 너무 긴 출력 로직

```
public void printModifyAttendanceResult(AttendanceHistory attendanceHistory,
        AttendanceHistory modifyAttendanceHistory) {
        LocalDateTime attendanceTime = attendanceHistory.getAttendanceTime().getAttendanceTime();
        LocalDateTime modifyAttendanceTime = modifyAttendanceHistory.getAttendanceTime()
            .getAttendanceTime();
        int month = attendanceTime.getMonthValue();
        int day = attendanceTime.getDayOfMonth();
        DayOfWeek dayOfWeek = DayOfWeek.calculateDayOfWeek(attendanceTime.toLocalDate());
        AttendanceType attendanceType = attendanceHistory.getAttendanceType();
        int hour = attendanceTime.getHour();
        int minute = attendanceTime.getMinute();

        int modifyHour = modifyAttendanceTime.getHour();
        int modifyMinute = modifyAttendanceTime.getMinute();
        AttendanceType modifyAttendanceType = modifyAttendanceHistory.getAttendanceType();

        System.out.println(MODIFY_ATTENDANCE_RESULT_MESSAGE.formatted(
            month, day, dayOfWeek.getName(), hour, minute, attendanceType.getName(),
            modifyHour, modifyMinute, modifyAttendanceType.getName())
        );
    }
```
출력의 일부 메서드 입니다. 
필요한 코드만 모아뒀다고 생각했지만 이번 미션의 규칙인 메서드 10라인을 훌쩍 넘어버렸습니다.

저는 객체의 필드값을 매개변수로 받는것 보다 객체 자체를 받아오는게 효율성 측면에서 좋다고 생각하고 있습니다.
ex)
필드값으로 받아오기
```
public void printAge(int age) {
        System.out.println(age);
    }
```

객체 자체를 받아오기
```
public void printAge(Crew crew) {
        System.out.println(crew.getAge());
    }
```

후자의 코드가 더 좋다고 판단한 이유는 필드값의 상태가 바뀌어도 ex) int -> long  처럼 타입이 변경되어도 따로 관리를 안 해줘도 되기 때문입니다.

- 1) 저의 생각이 맞는 생각인지 궁금합니다.
- 2) 맞다면 긴 출력 로직을 수정하기 힘든 상황이라고 판단하는데, 이 경우에 출력을 위한 DTO를 사용해도 될까요?
명확한 이유가 없다면 DTO 도입을 지양하고자 하는데, 이는 DTO 도입을 위한 충분한 근거일까요?
만약 이 또한 DTO를 도입할 명확한 이유가 아니라면, 어떤식으로 출력 메서드의 길이 조정을 할 수 있을지 궁금합니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+        int month = localDate.getMonthValue();
+        int day = localDate.getDayOfMonth();
+        for (int i = 1; i < day; i++) {
+            validateWeekdayAndCalculateAttendance(year, month, i);
```
여기서 year, month, day를 각각 넘기는 대신 LocalDate를 넘기면 파라미터 수를 많이 줄일 수 있어보이네요. 또, `LocalDate#withDayOfMonth`를 사용해도 좋을 것 같아요.

CodingMasterLSW replied:
수정해보겠습니다!

ddaaac commented on src/main/java/attendance/utils/AttendanceFileReader.java:
```diff
@@ -0,0 +1,40 @@
+            if (!crews.addCrew(crew)) {
+                crew = crews.findByCrewName(nickname);
+            }
+            crewAttendanceManager.addCrewAttendanceInfo(crew, attendanceHistory);
```
crewAttendanceManager는 도메인이고 AttendanceFilReader는 유틸성 클래스 같아요.

유틸 클래스에서 도메인을 직접 조작하는 것보다는 Map<Crew, History> 형식으로 읽어오는 것까지만 하는 건 어떨까요?

CodingMasterLSW replied:
동의합니다. 읽어서 다른 객체에서 조작하도록 리팩토링 진행하겠습니다

ddaaac commented on src/main/java/attendance/domain/AttendanceType.java:
```diff
@@ -0,0 +1,29 @@
+        this.decideValue = decideValue;
+    }
+
+    public static AttendanceType decideType(int value) {
```
value, decideType은 메서드가 하는 일을 직관적으로 나타내지 않는 것 같아요.
도메인 내용이 적당히 녹아들어있는 변수명을 사용해볼까요?

CodingMasterLSW replied:
value -> lateTime
decideType -> determineAttendanceTypeByLateTime
으로 수정해봤습니다.
lateTime은 괜찮은 것 같은데, determineAttendanceTypeByLateTime은 좀 긴 감이 있다고 생각이 드네요.

범블비가 봤을 때, 해당 메서드 네이밍은 명확한가요? 아니면 조금만 짧게 수정을 할까요?
명확하게 네이밍을 하려니 너무 길어지는 문제가 있는데, 적절한 기준이 있는지 궁금합니다!

ddaaac replied:
static 메서드는 시그니처에 클래스명이 포함되는 게 일반적입니다. 지금처럼 static import를 사용하지 않는다면
`AttendanceType.determineBy(lateTime)` 이렇게 사용한다면 충분히 자연스럽게 사용할 수 있겠죠.
아니면 `AttendanceType.of(lateTime)` 등을 활용할 수도 있겠구요.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+            return;
+        }
+        Optional<AttendanceHistory> hasDate = getAttendanceHistoryByDate(findDate);
+        calculateAttendance(hasDate, year, month, i);
```
Optional에 대한 처리를 밖에서 하면 Optional을 파라미터로 넘길 필요가 없겠네요.

CodingMasterLSW replied:
리팩토링 후 코드입니다.
```
Optional<AttendanceHistory> hasDate = getAttendanceHistoryByDate(calculateDay);
        if (hasDate.isEmpty()) {
            addAbsenceAttendance(calculateDay);
        }
```

```
private void addAbsenceAttendance(LocalDate calculateDay) {
        LocalDateTime notAttendanceTime = LocalDateTime.of(calculateDay, LocalTime.of(0, 0));
        AttendanceHistory attendanceHistory = AttendanceHistory.from(notAttendanceTime);
        addAttendanceHistory(attendanceHistory);
    }
```
내부에서 까거나, 밖에서 까거나 비슷하다고 생각하는데, Optional을 파라미터로 넘기면 좋지 않은 특별한 이유가 있는건가요?
아니면 단순히, 매개변수의 수를 줄이기 위함인가요??

ddaaac replied:
optional parameter로 검색하면 많이 나오는 내용입니다. 한 번 찾아보시면 좋을 것 같네요.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+        boolean isSame = attendanceHistories.stream()
+            .anyMatch(result -> result.getAttendanceTime().
+                getAttendanceTime().toLocalDate()
+                .isEqual(attendanceHistory.getAttendanceTime().getAttendanceTime().toLocalDate()));
```
`result.xxx(attendanceHistory)` 처럼 구현할 수 있겠네요. 
getter를 사용해서 두 값을 비교하는 로직은 항상 해당 객체에 메세지를 던지는 방식으로 리팩토링이 가능합니다.

CodingMasterLSW replied:
확인 후 수정했습니다.
. 하나 있을 때는 쉽게 수정했는데, 깊게 들어가니깐 복잡해지는 느낌이여서 나중에 고쳐야지... 하고 안 고치고 있었군요

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+        Optional<AttendanceHistory> attendanceHistory = getAttendanceHistoryByDate(
+            modifyDateTime.toLocalDate());
+        if (attendanceHistories.isEmpty()) {
+            throw new IllegalArgumentException(NOT_EXIST_ATTENDANCE.getMessage());
```
`Optional#orElseThrow`를 활용해볼까요?

CodingMasterLSW replied:
```
public AttendanceHistory modifyAttendanceResult(LocalDateTime modifyDateTime) {
        AttendanceHistory attendanceHistory = getAllAttendanceHistoryByDate(
            modifyDateTime.toLocalDate()).orElseThrow(
                () -> new IllegalArgumentException(NOT_EXIST_ATTENDANCE.getMessage()));
        AttendanceHistory modifyAttendanceHistory = AttendanceHistory.from(modifyDateTime);
        attendanceHistories.remove(attendanceHistory);
        attendanceHistories.add(modifyAttendanceHistory);
        return modifyAttendanceHistory;
    }
```

위와 같은 방식으로 수정했습니다만... 애초에 비어있을 때 예외가 아닌, Optional로 가져와서 추가 작업을 실행하는게 조금 어색하네요... 설계가 잘못된 것 같다는 느낌이 드는데, 제 생각이 맞을까요?

실제로 Optional을 가져와서 예외를 던지는게 아니라 비어있다면 비즈니스 로직을 실행하는 경우가 있는지 궁금합니다...!

ddaaac replied:
꼭 예외처리를 하기보다는 "비어있는 경우 아무 행동도 하지 않고 메서드를 종료한다" 정도는 있겠네요.
그런경우에는 예외를 try~catch 하기 보다는 orElseGet 등을 활용하는 게 좋을 것 같아요.
이번 로직에 그런 부분이 없다면 바로 throw 하는 것도 좋아보입니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+        for (AttendanceHistory attendanceHistory : attendanceHistories) {
+            AttendanceType attendanceType = attendanceHistory.getAttendanceType();
+            attendanceResult.put(attendanceType, attendanceResult.get(attendanceType) + 1);
+        }
```
`Collectors.groupingBy`, `Collectors.counting`을 활용해볼까요?

CodingMasterLSW replied:
처음 들어보는 메서드였는데, 범블비 덕분에 좋은 메서드를 알아갑니다!! 감사해요🙂

ddaaac commented on src/main/java/attendance/domain/AttendanceHistories.java:
```diff
@@ -0,0 +1,124 @@
+        return Collections.unmodifiableList(attendanceHistories);
+    }
+
+    public AttendanceHistory getValidationAttendanceDate(LocalDate localDate) {
```
여기도 메서드명을 보면 어떤 일을 하는지 쉽게 짐작이 가지 않네요

CodingMasterLSW replied:
조금 더 명확하게 수정해보겠습니다.

ddaaac commented on src/main/java/attendance/domain/CrewStatus.java:
```diff
@@ -0,0 +1,42 @@
+    public static CrewStatus calculateCrewStatus(Map<AttendanceType, Integer> attendanceResult) {
+        int validateValue = 0;
+        validateValue += attendanceResult.get(ABSENCE);
+        validateValue += attendanceResult.get(LATE) / 3;
```
`validateValue`를 불변 변수로 만들 수 있을 것 같네요.

CodingMasterLSW replied:
long validateValue = attendanceResult.get(ABSENCE) + attendanceResult.get(LATE) / 3;

이렇게 수정해봤습니다!

ddaaac commented on src/main/java/attendance/domain/CrewStatus.java:
```diff
@@ -0,0 +1,42 @@
+
+    CrewStatus(String name, int decideStatusValue) {
+        this.name = name;
+        this.decideStatusValue = decideStatusValue;
```
여기도 변수명을 조금 더 구체적으로 바꿔주세요.

CodingMasterLSW replied:
name -> statusDescription
decideStatusValue -> statusDecisionValue로 변경했습니다.

ddaaac commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,108 @@
+    private final Crews crews;
+    private final OutputView outputView;
+    private final CurrentDate currentDate;
+    private final CrewAttendanceManager crewAttendanceManager;
```
crews와 crewAttendanceManager가 따로 관리되는 이유가 있나요?

CodingMasterLSW replied:
1) 책임 분리
crewAttendanceManager가 담당하는 역할이 너무 많다고 생각했습니다. 
단순히 이름을 통해 크루를 찾아온다거나, 이런 역할은 crews로 따로 관리하면 책임을 줄일 수 있겠다고 생각해 분리시켰습니다.

2) 항상 keySet()을 가져와야 함

crews 분리 전
```
public Crew findByCrewName(String crewName) {
        Set<Crew> crews = crewAttendanceInfo.keySet();
        return crews.stream()
            .filter(crew -> crew.isSameCrew(crewName))
            .findAny()
            .orElseThrow(() -> new IllegalArgumentException(INVALID_CREW_NAME.getMessage()));
    }
```

crews 분리 후
```
public Crew findByCrewName(String name) {
        return crews.stream()
            .filter(crew -> crew.isSameCrew(name))
            .findAny()
            .orElseThrow(() -> new IllegalArgumentException(INVALID_CREW_NAME.getMessage()));
    }

```

findByCrewName의 경우, 자주 사용되는 호출이라고 생각했습니다.
그런 측면에서 메서드를 한 줄이라도 줄이는게 효율적이지 않을까 생각했습니다.

위와 같은 생각으로 Crews, CrewAttendanceManager를 따로 관리하고 있습니다.
다만, CrewAttendanceManger에서 
 Map<Crew, AttendanceHistory>  crewAttendanceInfo 를 가지고 있다는 건, 충분히 해당 객체에서 crews의 역할을 대체할 수 있습니다.

이 부분에 대해 범블비의 의견이 궁금합니다. 만약 범블비라면, crews 객체를 삭제하실건가요? 아니면 위의 의견이 합당하다고 생각해 그대로 두실건가요?

ddaaac replied:
같은 데이터를 여러 객체에서 관리하면 어떤 일이 벌어질지 생각해보시면 좋을 것 같습니다.
누군가 코드를 작성하다가 실수로 CrewAttendanceManager에는 Crew를 추가하는 코드를 구현했는데 Crews에 추가를 하지 않으면 어떻게 될까요?

ddaaac commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,108 @@
+    private void checkDangerousCrews() {
+        DangerousCrews dangerousCrews = DangerousCrews.create();
+        List<DangerousCrew> foundDangerousCrewResults = dangerousCrews.findDangerousCrewsAndSort(
+            crewAttendanceManager, crews);
```
`crewAttendanceManager.findDangerousCrew`와는 어떤 차이가 있을까요?

CodingMasterLSW replied:
crewAttendanceManager.findDangerousCrew 메서드는 발견하지 못 했습니다...
혹 어떤 코드를 말씀하시는건지 알 수 있을까요?

ddaaac replied:
저렇게 구현했을 떄의 차이점에 대해 생각해보시면 좋을 것 같아서 남긴 코멘트입니다.

ddaaac commented on src/main/java/attendance/domain/CrewStatus.java:
```diff
@@ -0,0 +1,24 @@
+package attendance.domain;
+
+public enum CrewStatus {
```
> 다만, crew 객체의 역할이 조금 모호하다고 생각해 crewStatus를 crew객체에 포함시킬까 고민중입니다.
이 경우에, crewStatus의 default 값을로 CrewStatus.CLEAR로 설정하고, 이후에 크루 상태를 변경하려고 합니다.
이런 리팩토링 사항에 대해 범블비의 의견이 궁금합니다


음 현재 구조에서는 따로 두는 게 나아보이긴 하네요. 
crew가 갖고 있게 하면 출석을 추가/변경할때마다 계산을 새로해줘야할 것 같아요.

ddaaac commented on src/main/java/attendance/domain/DangerousCrews.java:
```diff
@@ -0,0 +1,48 @@
+        return new DangerousCrews();
+    }
+
+    public List<DangerousCrew> findDangerousCrewsAndSort(
```
> 기존에는 view에서 제적위험자를 정렬하고 있었습니다. 이를 수정하기 위해 controller에서 정렬을 시도했었고,
controller에서 비즈니스 로직을 수행하는건 올바르지 않다 생각해 DangerousCrew, DangerousCrews 객체 생성 후 책임을 전가했습니다.
범블비가 봤을 때, DangerousCrew는 필요한 객체라고 생각하시나요? 아니면 과한 분리라고 생각하시나요?? 의견이 궁금합니다.

말씀하신 DangerousCrews를 만들기 위한 과정이 타당해보이네요. 충분히 있을 법한 객체라 생각됩니다. 테스트도 용이하게 할 수 있구요.

ddaaac commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,113 @@
+    private static final String INTERVIEW_TARGET_MESSAGE = "면담 대상자입니다.";
+    private static final String DANGEROUS_CREW_INFO_MESSAGE = "제적 위험자 조회 결과";
+    private static final String DANGEROUS_CREW_INFO = "- %s: 결석 %s회 지각 %s회 (%s)";
+
+    public void printModifyAttendanceResult(AttendanceHistory attendanceHistory,
```
> 후자의 코드가 더 좋다고 판단한 이유는 필드값의 상태가 바뀌어도 ex) int -> long 처럼 타입이 변경되어도 따로 관리를 안 해줘도 되기 때문입니다.
저의 생각이 맞는 생각인지 궁금합니다.
맞다면 긴 출력 로직을 수정하기 힘든 상황이라고 판단하는데, 이 경우에 출력을 위한 DTO를 사용해도 될까요?
명확한 이유가 없다면 DTO 도입을 지양하고자 하는데, 이는 DTO 도입을 위한 충분한 근거일까요?
만약 이 또한 DTO를 도입할 명확한 이유가 아니라면, 어떤식으로 출력 메서드의 길이 조정을 할 수 있을지 궁금합니다.


DTO의 목적이 출력 메서드의 길이를 줄이기위함만은 아니겠지만, 젠슨이 생각하신 방향이 DTO 도입의 충분한 이유가 되는 걸로 보이네요. 결국에 객체가 가진 정보를 View에서 보여줘야하는데 객체 자체를 넘겼을 때 View에서 도메인 로직을 호출할 수 있으니(=View와 도메인의 결합이 생기니) 객체가 가진 정보만을 갖는 클래스를 따로 만들자에서 DTO가 나왔다고 볼 수도 있으니깐요.

객체는 로직을 수행하고 DTO는 보여줘야할 값에 대한 정보를 들고 있는 것 역시 명확한 책임 분리라고 할 수도 있구요.

이와 별개로 메서드 라인은 그냥 메서드 분리를 통해서도 관리할 수 있을 것 같네요.

CodingMasterLSW replied:
확인했습니다!

ddaaac reviewed (changes requested):
안녕하세요 젠슨 👋 

리뷰 반영 잘 해주셨습니다!

몇 가지 코멘트 남겼으니 확인해주시고, 추가로 궁금한 점 있으시면 남겨주세요.

CodingMasterLSW committed:
refacrtor: AttendanceHistories 리팩토링 진행

함수명, 매개변수 명을 보다 직관적으로 변경하였고, for문을 stream으로 변경하였음.

CodingMasterLSW committed:
rename: AttendanceType, CrewStatus의 메서드, 필드값 명확하게 변경

CodingMasterLSW committed:
refactor: CrewStatus validateValue 계산 식 변경

CodingMasterLSW committed:
refactor: FileReader와 도메인 로직 분리

단순히 값을 읽도록 리팩토링을 진행하였음

CodingMasterLSW committed:
refactor: outputView 메서드 분리

CodingMasterLSW commented:
범블비의 피드백을 토대로 수정해봤습니다!
리팩토링을 진행하며 궁금한 점은 댓글 달아놨습니다!

ddaaac reviewed (approved):
안녕하세요 젠슨 👋 

리뷰 반영 잘 해주셨습니다.

몇 가지 코멘트 남겼으니 다음 단계 진행하면서 참고해주세요
